community/kernel/src/main/java/org/neo4j/kernel/impl/api/index/IndexingService.java

```diff
@@ -545,28 +547,29 @@ private void applyRecoveredUpdates() throws IOException, IndexEntryConflictExcep
                     updater.remove( recoveredNodeIds );
                 }
 
-                Iterable<NodeUpdates> updates = readNodeUpdatesFromRecoveredStore();
-                convertToIndexUpdatesAndApply( updates, IndexUpdateMode.RECOVERY );
-                monitor.appliedRecoveredData( updates );
+                Iterator<NodeUpdates> updates = recoveredNodeUpdatesIterator();
+                Iterator<IndexEntryUpdate<LabelSchemaDescriptor>> indexEntryUpdates =
+                        Iterators.flatMap( nodeUpdates -> convertToIndexUpdates( nodeUpdates ).iterator(), updates );
+                while ( indexEntryUpdates.hasNext() )
+                {
+                    IndexEntryUpdate<LabelSchemaDescriptor> indexUpdate = indexEntryUpdates.next();
+                    monitor.applyingRecoveredUpdate( indexUpdate );
+                    IndexUpdater updater = updaterMap.getUpdater( indexUpdate.indexKey().schema() );
+                    if ( updater != null )
+                    {
+                        updater.process( indexUpdate );
+                    }
```

Comment:
This code is also used in `apply( Iterable<IndexEntryUpdate<LabelSchemaDescriptor>> updates, IndexUpdateMode updateMode )`. Can we extract it?
```
IndexUpdater updater = updaterMap.getUpdater( indexUpdate.indexKey().schema() );
if ( updater != null )
{
    updater.process( indexUpdate );
}
```